### ISSUE TITLE
The number base in RSA used for 64 bit architecture

### ISSUE THREAD

**Issue by lixinyu2016** [milagro-crypto/amcl#66]
In the "amcl/version3/c/config64.py" code ([https://github.com/milagro-crypto/amcl/blob/master/version3/c/config64.py](url))，the RSA parameter is 

# rsaset(big,ring,big_length_bytes,bits_in_base,multiplier)
For example, 
                #rsaset("1024","2048","128","58","2")
		#rsaset("512","2048","64","60","4")
                #rsaset("384","3072","48","56","8")
		#rsaset("256","2048","32","56","8")

Question 1: Is the length of the number base for 64-bit architecture, namely "bits_in_base" fixed for a specific "big"?

Question 2: As said in the config64.py, "#256 is slower but may allow reuse of 256-bit BIGs used for elliptic curve, #512 is faster.. but best is 1024", then my understanidng is: for one RSA test, the parameter "big" would influence the final efficiency, and the larger the "big" is , the faster for the test, right?

Question 3: we want to test one RSA-based scheme, for example, we want to test the timing cost for one exponentiation in Z_N, but the  exponent  is not eaual to that of the public RSA encryption key "e" (for example, e=65537 ), is it feasible by using this library?

**Comment by lixinyu2016** [comment]
In addition, if we want to consider a 7680-bit or a 15360-bit N for 192-bit and 256-bit security RSA, then how large the parameter "big" as well as the " bits_in_bases" should we select?

**Comment by mcarrickscott** [comment]
2048-bit RSA requires n=p.q where p and q are 1024-bits, and n is 2048
bits. This may be represented internally in many ways. For example

#rsaset("512","2048","64","60","4")

Here p and q are each 1024 bits, and are each constructed from two 512-bit
(=64 byte) BIG numbers. Each digit of a BIG number stores 60 bits, so 9
digits are required per BIG. A 2048-bit number is represented as 4  BIGS.

Q1: Its not fixed, but as recommended by the check.cpp utility. For example
check 64 512 outputs

Solution for BASEBITS= 60, Words Per Big=9, Max excess= 28
Solution for BASEBITS= 58, Words Per Big=9, Max excess= 10
Solution for BASEBITS= 56, Words Per Big=10, Max excess= 30

Q2: Internally Karatsuba's recursive method is used. So p could be
represented by 4x256, or 2x512, or 1x1024. There will be an optimal sweet
spot, which can only be determined by experimentation. It may not
necessarily be the largest, and indeed will not be for much bigger moduli.

Q3: The public exponent 65537 is most commonly used, and recommended by
standards. But any other odd prime can be used, see the rsa-2048() function
in testall.c

Q4: We haven't tried using N greater than 4096 bits, as it has not been
requested.  Note that the library is mainly intended to support elliptic
curves.

Mike

**Comment by lixinyu2016** [comment]
@mcarrickscott

Thank you for your response. Now we want to evaluate our pairing-based scheme  with other schemes based on RSA (in particular, the GQ signature scheme) in different security levels, for example, 128-bit and 192-bit security. For 192-security, can we reset  the RSA parameters, for example: #rsaset("960","7680","120","60","8"), to test the 7680-bit RSA?

**Comment by mcarrickscott** [comment]
Yes, that works fine. I tested it with the 64-bit c++ version, using

rsaset("960","RSA7680","120","56","8")

Then edited benchtest_all.cpp to globally replace "2048" with "7680"

Then I also had to edit the file rsa_support.h, to change 512 to 960

#define MAX_RSA_BYTES 960 // Maximum..

Finally built the library and executed the benchtest_all program, which
output

Testing/Timing 7680-bit RSA
Generating 7680-bit RSA public/private key pair
RSA gen -       10 iterations   45159.38 ms per iteration
RSA enc -     2206 iterations       4.53 ms per iteration
RSA dec -       63 iterations     160.96 ms per iteration
All tests pass

The check utility output

Not enough Fp excess for BASEBITS= 60
Solution for BASEBITS= 58, Words Per Big=17, Max excess= 26
Solution for BASEBITS= 56, Words Per Big=18, Max excess= 30

So other options that might be a little quicker would include

rsaset("960","RSA7680","120","58","8")
rsaset("480","RSA7680","120","56","16")
rsaset("1920","RSA7680","120","56","4")
rsaset("480","RSA7680","120","58","16")
rsaset("1920","RSA7680","120","58","4")


Mike

**Comment by lixinyu2016** [comment]
@mcarrickscott 

Thank you very much for your detailed responses. We would perform our test later by the method you suggested, and may need your further suggestions if we meet some other problems, thank you in advance!

**Comment by lixinyu2016** [comment]
@mcarrickscott
Hi, as you said before, 

Q3: The public exponent 65537 is most commonly used, and recommended by
standards. But any other odd prime can be used, see the rsa-2048() function
in testall.c

During our test for #rsaset("384","3072","48","56","8"), we found that the big number "384" is fixed , how can we select other public exponent e? For example, the length of e  we need is 128 bit, how can we use a 384-bit big number to represnt e.

Our way is to obtain 128-bit e by  right shifting 256 bit of a 384-bit number, however, by this method, the timing cost for one exponentiation m^e mod N would be very large (about  458ms), while when we use Miracl library,  the time is only about 3.7 ms.  Below is our test code, is it right?

![catch 01-08-19-42-07](https://user-images.githubusercontent.com/37723348/50828861-87b2ea00-137d-11e9-8661-84b59e11a367.jpg)

**Comment by mcarrickscott** [comment]
Use of such an exponent was not considered in the design of the library..

First of all only odd exponents are supported (otherwise its not RSA!)

A good idea would be to use FF_3072_sk*s*pow(result,x,e[0],N,8) instead of
FF_3072_skpow(). That should be 8 times faster

A better idea for a 128-bit exponent would be to introduce a new powering
function the same as FF_skspow() but with

    for (i=8*MODBYTES_XXX-1; i>=0; i--)

replaced with

    for (i=BIG_XXX_nbits(e)-1; i>=0; i--)


That should be another 3 times faster. The best idea (since operations with
e use all public values, and hence do not need to be side-channel
resistant), would be to modify a simple square-and-multiply algorithm as
used in FF_pow(), but with int e replaced with dchunk e. Note that a dchunk
should be 128-bits for a 64-bit build of the library, and capable of
storing your exponent e

Mike

**Comment by lixinyu2016** [comment]
@mcarrickscott

Thank you for that. In addition , is the way we obtain a 128-bit number from a 384-bit big number correct? namely,   by right shifting 256 bit of a 384-bit number.

**Comment by lixinyu2016** [comment]
@mcarrickscott 

We tried the test usign the optimization measure and it performed pretty  well, thank you for that.

By the way, in the benchark, how much is the curve equation for BLS 12, BLS 24, namely the "b" in y^2 = x^3 +b. According to [https://eprint.iacr.org/2017/334.pdf](url), b= -2 for BLS 12 BLS 24, is it also the case in this benchmark?

**Comment by mcarrickscott** [comment]
The curves supported in the AMCL library are not the same as those in
https://eprint.iacr.org/2017/334.pdf <http://url/>

The b parameter for each curve can be found in the relevant ROM file, for
example in rom_curve_BLS24.c

const int CURVE_B_I_BLS24= 19;

So the b parameter is 19 for this curve.

Mike

**Comment by lixinyu2016** [comment]
Our problem has been solved, and thank you very much!
